feat(importer): build latest config from data

Y4nnikH opened the pull request:
closes #3566 

#### Breaking Change for developers
When doing an upgrade from a previous importer-rework status, the primary key of table latest_config needs to be updated by hand to mgm_id

Y4nnikH committed:
feat(importer): check latest config

Y4nnikH committed:
feat(db): change latest config key to mgm_id

Y4nnikH committed:
wip: build normalized config from data

Y4nnikH committed:
Merge remote-tracking branch 'CactuseSecurity/importer-rework' into fix/latest-config-in-db

tpurschke commented on roles/database/files/sql/creation/fworch-create-constraints.sql:
```diff
@@ -14,7 +14,6 @@ Alter Table "changelog_service" add Constraint "alt_key_changelog_service" UNIQU
 Alter Table "changelog_user" add Constraint "alt_key_changelog_user" UNIQUE ("abs_change_id");
 Alter Table "import_changelog" add Constraint "Alter_Key14" UNIQUE ("import_changelog_nr","control_id");
 Alter Table "import_control" add Constraint "control_id_stop_time_unique" UNIQUE ("stop_time","control_id");
-Alter table "latest_config" add CONSTRAINT unique_latest_config_mgm_id UNIQUE ("mgm_id");
```
why have you deleted this constraint?

tpurschke replied:
I see, you made it the primary key.

tpurschke commented on roles/lib/files/FWO.Data/Device.cs:
```diff
@@ -39,8 +39,8 @@ public class Device
 
-        // [JsonProperty("rulebase_links"), JsonPropertyName("rulebase_links")]
-        // public RulebasePerGateway[] Rulebases { get; set; } = [];
+        [JsonProperty("rulebase_links"), JsonPropertyName("rulebase_links")]
```
it feels very strange that we did not have the rulebase_links under device - how was the reporting working up to now?! @Robin-Smets any thoughts on this?

Robin-Smets replied:
There are extra model classes for reporting. In ManagementReport you have a property rulebases, that contains all rulebases and a property Devices, which has the property RulebaseLinks. In ReportData you have a property ManagementData, that contains all fetched ManagementReport objects... But you are totally right, the structure is very confusing. Thats why I created some other properties during my work on the reports. I suggest we talk about that in depth on the next cactus day.

tpurschke commented on roles/lib/files/FWO.Data/Middleware/NormalizedConfigParameters.cs:
```diff
@@ -0,0 +1,15 @@
+        [JsonProperty("mgm-ids")]
+        public int[] ManagementIds { get; set; } = [];
+
+        [JsonProperty("config=time")]
```
config=time looks weird - can you explain this to me?

Y4nnikH replied:
fixed this typo

tpurschke reviewed (commented):
looking nice
some questions comments, see inline

Y4nnikH committed:
Merge remote-tracking branch 'CactuseSecurity/importer-rework' into fix/latest-config-in-db

Y4nnikH committed:
feat: build normalized config from database

Y4nnikH committed:
Merge remote-tracking branch 'CactuseSecurity/importer-rework' into fix/latest-config-in-db

Y4nnikH committed:
fix(db): allow importer to access uiuser

Y4nnikH committed:
refactor: fixes and satisfy sonarqube

Y4nnikH committed:
Merge remote-tracking branch 'CactuseSecurity/importer-rework' into fix/latest-config-in-db

Y4nnikH committed:
refactor: sq

Y4nnikH committed:
refactor: update TODO comments and improve error handling in NormalizedConfigController

Y4nnikH commented on roles/importer/files/importer/model_controllers/fwconfig_import.py:
```diff
@@ -80,20 +71,33 @@ def import_management_set(self, import_state: ImportStateController, service_pro
+        mgm_id = self.import_state.lookupManagementId(manager.ManagerUid)
+        if mgm_id is None:
+            raise FwoImporterError(f"could not find manager id in DB for UID {manager.ManagerUid}")
+        #TODO: clean separation between values relevant for all managers and those only relevant for specific managers
```
see #3646

Y4nnikH commented:
> ## [![Quality Gate Failed](https://camo.githubusercontent.com/ce402bcc4cf818eda660128f0b27a891eb8c9f4e0d783303ea667cb0d4b07f95/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f5175616c6974794761746542616467652f71672d6661696c65642d323070782e706e67)](https://sonarcloud.io/dashboard?id=CactuseSecurity_firewall-orchestrator&pullRequest=3616) **Quality Gate failed**
> 
> Failed conditions ![](https://camo.githubusercontent.com/bd1d5234b37410841ad8cd9d85c84eeecaec92ae18a95c926ae3a59a17c29a65/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6661696c65642d313670782e706e67) [1 New issue](https://sonarcloud.io/project/issues?id=CactuseSecurity_firewall-orchestrator&pullRequest=3616&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true)
> 
> [See analysis details on SonarQube Cloud](https://sonarcloud.io/dashboard?id=CactuseSecurity_firewall-orchestrator&pullRequest=3616)
> 
> ![](https://camo.githubusercontent.com/0d8093b061da31b2dde5907c87c133451a9fed9ab208145ec3750960f13f438a/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6c696768745f62756c622d313670782e706e67) Catch issues before they fail your Quality Gate with our IDE extension ![](https://camo.githubusercontent.com/cf28d9441c2be84b83964b59542f8523d293ea1d96a9d3cea77e7962537f0175/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f736f6e61726c696e742d313670782e706e67) [SonarQube for IDE](https://www.sonarsource.com/products/sonarlint/features/connected-mode/?referrer=pull-request)

I disagree with sonarqube here. Splitting the recursive function into multiple functions that then call back to the main function for recursiveness seems more complex/confusing to me.

alf-cactus commented on roles/importer/files/importer/fortiadom5ff/fmgr_service.py:
```diff
@@ -39,7 +39,7 @@ def normalize_service_object(obj_orig, svc_objects):
         name = str(obj_orig['name'])
 
-    color = str(obj_orig.get('color', 1))
+    color = 'foreground' #TODO: color mapping. what is color: 0? (nativeconfig entwickler_fortimanager_stand_2025-07-27, service object 'gALL')
```
I fixed this in my branch by defaulting all uncolored objects to 'black'

alf-cactus reviewed (approved):
The forti changes could cause some mild merge conflicts with my branch, but nothing that will be to complicated to solve on my end

alf-cactus reviewed (dismissed):
When testing a initial fortinet import I get following error:
fwo_api: file not found: 7usr/local/fworch/fwo-api-calls/import/getLastSuccessImport.graphql

Y4nnikH committed:
Merge remote-tracking branch 'CactuseSecurity/importer-rework' into fix/latest-config-in-db

Y4nnikH committed:
feat(importer): always check for import db consistency

Y4nnikH committed:
fix(importer): correctly set global config

also add comments for structural rework later

Y4nnikH commented:
```
RUNNING HANDLER [tests-integration : delete ldap test data] ********************
changed: [localhost] => (item=uid=user2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP,ou=tenant2_test_YWsEcsHdIa GlgwOzdedjvQoEVTxceDjP,ou=operator,ou=user,dc=fworch,dc=internalobjectClass: topobjectClass: inetOrgPersoncn: user2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjPgivenName: user2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjPsn: RotuserPassword:: VVRQS0xySnlaSkRLb2ZUTElkeU9lQWJidEVvVmxEVE4=uid: user2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP# group1_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP, group, fworch.internal)
changed: [localhost] => (item=uid=user1_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP,ou=tenant1_test_YWsEcsHdIa GlgwOzdedjvQoEVTxceDjP,ou=operator,ou=user,dc=fworch,dc=internalobjectClass: topobjectClass: inetOrgPersoncn: user1_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjPgivenName: user1_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjPsn: GelbuserPassword:: WXRvbmhRUVNYRHJXUHpJTnRZZ3FjVm14bGpNcWh4REI=uid: user1_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP# user2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP, tenant2_test_YWsEcsHdIaGlgwOzded jvQoEVTxceDjP, operator, user, fworch.internal)
changed: [localhost] => (item=ou=tenant2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP,ou=operator,ou=user,dc=fw orch,dc=internalobjectClass: topobjectClass: organizationalUnitou: tenant2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP# user1_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP, tenant1_test_YWsEcsHdIaGlgwOzded jvQoEVTxceDjP, operator, user, fworch.internal)
changed: [localhost] => (item=ou=tenant1_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP,ou=operator,ou=user,dc=fw orch,dc=internalobjectClass: topobjectClass: organizationalUnitou: tenant1_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP# tenant2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP, operator, user, fworch.interna l)
changed: [localhost] => (item=cn=ownergroup_F_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP,ou=group,dc=fworch,d c=internalobjectClass: topobjectClass: groupOfUniqueNamescn: group2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjPcn: ownergroup_F_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjPbusinessCategory: ownergroupuniqueMember:# search resultsearch: 2result: 0 Success# numResponses: 9# numEntries: 8)
changed: [localhost] => (item=cn=ownergroup_D_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP,ou=group,dc=fworch,d c=internalobjectClass: topobjectClass: groupOfUniqueNamescn: group2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjPcn: ownergroup_D_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjPbusinessCategory: ownergroupuniqueMember:# ownergroup_F_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP, group, fworch.internal)
changed: [localhost] => (item=cn=group2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP,ou=group,dc=fworch,dc=inte rnalobjectClass: topobjectClass: groupOfUniqueNamescn: group2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjPuniqueMember:# ownergroup_D_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP, group, fworch.internal)
changed: [localhost] => (item=cn=group1_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP,ou=group,dc=fworch,dc=inte rnalobjectClass: topobjectClass: groupOfUniqueNamescn: group1_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjPuniqueMember:# group2_test_YWsEcsHdIaGlgwOzdedjvQoEVTxceDjP, group, fworch.internal)
```

Any idea why the test install failed? @tpurschke
